### ISSUE TITLE
Precedence between Tokenization Rules

### ISSUE THREAD

**Issue by caprice-j** [systemincloud/rly#1]
Hi, thank you for providing a great library. 

I'm now trying to implement my language by rly.
Is there any way to prefer one tokenisation rule in Lexer to the other rules?

Specifically, I have the following two rules in R6::R6Class("Lexer"):

t_THEME = '(\\+|\\|)\\s*theme',
t_LAYER = function(re='(\\+|\\|)\\s*[a-z_]+', t) { ... }

THEME is a token which is preceded with whitspaces and a plus sign (ex " + theme"),
and LAYER is a similar token whose suffix is arbitrary alphanumeric names (ex " + myvar1" or " + ididid").

Currently my lexer wrongly recognizes `rly::lex(module=myLexer)$input(' + theme')` as LAYER.

I know if I add the negation of 'theme' in LAYER I can achieve what I'd like to do, but I guess there is an interface to define precedence of tokenization rules.

**Comment by marekjagielski** [comment]
Hi,
Great that you like the library. I need to point out that the library is recently written and probably with some issues. I am just now using it for my first project.
Rly is highly based on ply python library. In case of ply, the author introduced the order as a determiner of the priority of rules in case of multiple matches.

Unfortunately, I have to say that today with rly you can't determine priority of rules. R6 class is an environment type. When implementating, I couldn't determine the order of the methods declared in R6 class. So today you have to be specific with your regex. 

I am concerned about this limitation and I would like to solve this issue maybe by adding an optional attribute e.g. 'priority'.

**Comment by marekjagielski** [comment]
To respond to this issue Ply library is doing sorting in get_rules method:

>         # Sort the functions by line number
>         for f in self.funcsym.values():
>             f.sort(key=lambda x: x[1].__code__.co_firstlineno)

>        # Sort the strings by regular expression length
>        for s in self.strsym.values():
>            s.sort(key=lambda x: len(x[1]), reverse=True)

**Comment by caprice-j** [comment]
Thank you for your quick and detailed reply.

I understand that this library is still in a very early stage (just a few month old?).
You don't have to hurry to deal with my issues. I found that rly can already deal with the precedence feature in the following way and I'm fully satisfied with this.

```
    t_LAYER = function(re='(\\+|\\|)\\s*[a-z_]+', t) {
                        if (grepl('(\\+|\\|)\\s*theme', t$value)) {
                            t$type <- 'THEME'
                            return(t)
                        }
                
```

As you said it might be better if you can add a specific way to do that. In my humble opinion when I heard rly is a clone of ply, I also assumed that most of the usage would be the same.

Thus it might be great if you can solve this priority issue in the same way as ply (line number and sorting).

Actually, I first tried to change the order of function definitions in Lexer hoping to change the precedence. I haven't used ply before so novice users might behave in the same way. Though I don't know how to get the line numbers of R6 definitions.

**Comment by marekjagielski** [comment]
Hi,
I made an analyse of today's solution. To get symbols from R6 class I use:

>       tsymbols <- c(grep('^t_', names(self$module$public_fields), value=TRUE),
>                     grep('^t_', names(self$module$public_methods), value=TRUE))

Surprisingly 'names' function on public_fields and public_methods keeps the order of appearing in public of R6 class. When doing the matching, first, I check methods then fields, so the functions have always higher priority. Inside group of functions or group of fields it seems that order counts.

This solution is close to ply. In ply functions also have higher priority. Then, the order counts among functions. However inside fields, the length of regex is determiner (differently to rly).

From: http://www.dabeaz.com/ply/ply.html

> When building the master regular expression, rules are added in the following order:
> 1. All tokens defined by functions are added in the same order as they appear in the lexer file.
> 2. Tokens defined by strings are added next by sorting them in order of decreasing regular expression length (longer expressions are added first).

I think that order of fields is more intuitive anyway. So I would not change the current solution.

I added tests showing issue: https://github.com/systemincloud/rly/blob/master/tests/testthat/test.lex_precedence.R
